File: crates/reincarnate-core/src/system/timing.rs
/// Timing system trait — handles frame pacing and time tracking.
pub trait Timing {
    /// Seconds elapsed since last frame.
    fn delta_time(&self) -> f64;

    /// Total seconds elapsed since start.
    fn elapsed(&self) -> f64;

    /// Current frame number.
    fn frame_count(&self) -> u64;

    /// Target frames per second (0 = uncapped).
    fn target_fps(&self) -> u32;

    /// Set target frames per second.
    fn set_target_fps(&mut self, fps: u32);

    /// Called once per frame to advance timing state.
    fn tick(&mut self);
}

File: runtime/flash/ts/timing.ts
/** Flash timing shim — delta time, frame counter, target FPS. */

export class TimingShim {
  private _lastTime = performance.now() / 1000;
  private _startTime = performance.now() / 1000;
  private _dt = 0;
  private _frames = 0;
  private _fps = 60;

  delta_time(): number { return this._dt; }
  elapsed(): number { return performance.now() / 1000 - this._startTime; }
  frame_count(): number { return this._frames; }
  target_fps(): number { return this._fps; }
  set_target_fps(value: number): void { this._fps = value; }

  tick(): void {
    const now = performance.now() / 1000;
    this._dt = now - this._lastTime;
    this._lastTime = now;
    this._frames++;
  }
}
